File: backend/src/main/java/com/trade/service/OrderService.java
package com.trade.service;

import com.trade.dto.order.OrderRequest;
import com.trade.dto.order.OrderResponse;

import java.util.List;

/**
 * Contract for Export Shipment Request management.
 *
 * Exporters: create, update, view their own requests.
 * Logistics: view pending requests, accept, reject.
 */
public interface OrderService {

    // ── Exporter operations ──────────────────────────────────────────────────

    /** Returns all requests created by this exporter. */
    List<OrderResponse> getAllOrdersForExporter(String exporterEmail);

    /** Returns a single request — exporter must own it. */
    OrderResponse getOrderById(Long id, String exporterEmail);

    /** Creates a new shipment request with status PENDING_LOGISTICS. */
    OrderResponse createOrder(OrderRequest request, String exporterEmail);

    /** Updates a request — only allowed while status is PENDING_LOGISTICS. */
    OrderResponse updateOrder(Long id, OrderRequest request, String exporterEmail);

    // ── Logistics operations ─────────────────────────────────────────────────

    /** Returns all requests with status PENDING_LOGISTICS, excluding those already rejected by this partner. */
    List<OrderResponse> getPendingRequests(String logisticsPartnerEmail);

    /**
     * Logistics partner accepts a request.
     * Changes status to LOGISTICS_ACCEPTED, saves partner, auto-creates Shipment.
     */
    OrderResponse acceptRequest(Long id, String logisticsPartnerEmail);

    /**
     * Logistics partner rejects a request.
     * Records the rejection. Request stays PENDING_LOGISTICS for other partners.
     * If all registered logistics partners have rejected, status becomes REJECTED.
     */
    OrderResponse rejectRequest(Long id, String logisticsPartnerEmail);
}

File: frontend/src/services/orderService.js
/**
 * orderService.js — Orders API (shared by Exporter and Logistics roles).
 *
 * Backend: OrderController @ /api/orders
 */

import { get, post, put, patch } from './httpClient';

export const ordersApi = {
  // ── Exporter endpoints ─────────────────────────────────────────────────────

  /**
   * GET /api/orders — list exporter's own orders
   */
  getMyOrders() {
    return get('/orders');
  },

  /**
   * GET /api/orders/:id
   */
  getById(id) {
    return get(`/orders/${id}`);
  },

  /**
   * POST /api/orders — create a new shipment request
   * @param {object} order - { productId, destinationCountryId, quantity, ... }
   */
  create(order) {
    return post('/orders', order);
  },

  /**
   * PUT /api/orders/:id — update (only while PENDING_LOGISTICS)
   */
  update(id, order) {
    return put(`/orders/${id}`, order);
  },

  // ── Logistics endpoints ────────────────────────────────────────────────────

  /**
   * GET /api/orders/pending — all PENDING_LOGISTICS requests (excluding own rejections)
   */
  getPending() {
    return get('/orders/pending');
  },

  /**
   * PATCH /api/orders/:id/accept — accept a shipment request
   */
  accept(id) {
    return patch(`/orders/${id}/accept`);
  },

  /**
   * PATCH /api/orders/:id/reject — reject a shipment request
   */
  reject(id) {
    return patch(`/orders/${id}/reject`);
  },
};
